### PR TITLE
Feature/jsUrl Rev

### LINKED ISSUE #369: app method url() should have an argument indicating a regular or ajax url
Some integrations (like ibelar/atk-wordpress) need to differentiate between ajax and regular requests when creating a url. MasterCrud does a lot with regular requests. (quickSearch as well I believe)

I suggest the url interface to be:
```php
/**
     * @param bool $ajax
     * @return string
     */
    public function url($ajax = true)
```
@ibelar, do you agree?

### PR DESCRIPTION
**Pull request by ibelar**
Revision using App:jsUrl() without view property.

resolves #369

### CODE REVIEW

**Comment by ibelar** [comment]
@romaninsh - Here is another example of using jsUrl() method vs needAjax property.

When using Tab with VirtualPage in WP, generating url would need to be done via jsUrl() in order to work properly in WP. Tabs::AddTab() method is easy to update for VirtualPage to generate jsUrl() when using needAjax property. See commit [5bcd1b2](https://github.com/atk4/ui/pull/372/commits/5bcd1b270dab57ad9abdb2fd0f38519c3d5c5f4f) in atk4\ui#372.

Not sure how to do so using jsUrl() without altering VirtualPage. Let me know your thoughts.

**Comment by ibelar** [comment]
@romaninsh - Never mind my last (delete) comments. 

Manage to make it work by using a new property on Callback ($needJsUrl)  instead of View. 
Now, callback will call proper owner url or jsUrl method according to this new property. 
This way, Tabs and jsModal are easy to modify and a lot clearer while jsCallback just need a property value change to generate jsUrl.

Hope this is ok. Let me know your thoughts.

File: src/Tabs.php [hunk 1]
```diff
@@ -15,12 +15,13 @@ class Tabs extends View
...
-     * @param mixed $name   Name of tab or Tab object
-     * @param mixed $action Optional callback action or URL (or array with url + parameters)
+     * @param mixed $name      Name of tab or Tab object
+     * @param mixed $action    Optional callback action or URL (or array with url + parameters)
+     * @param bool  $needJsUrl Whether the virtual page should generate a jsUrl or not.
      *
      * @return View
      */
-    public function addTab($name = null, $action = null)
+    public function addTab($name = null, $action = null, $needJsUrl = false)
```
Comment (romaninsh) [inline src/Tabs.php]: how do i know if virtual page should generate jsURL or not?

**Comment by romaninsh** [comment]
Yes, this looks much cleaner. I still don't understand why i need to pass extra argument to addTab though, can you clarify?

**Comment by romaninsh** [comment]
OH one note - we need a paragraph in the docs regarding jsURL. What's the policy on the uppercase ? should we use jsURL? I think we uppercase acronyms entirely (getURL) unless they are in the beginning (dbConnect).

**Comment by ibelar** [comment]
In regards to extra argument for addTab. When using a regular url to load tab in WP, this will output the entire page instead of the tab portion needed, so it must use an ajax request which is route via jsUrl. It's really needed for WP to output tab dynamically.

I will update PR to us jsURL instead.

**Comment by romaninsh** [comment]
Oh i think we shouldn't be using regular URL in tab. I know one of the tests may be using that, but I am not happy about it. It's either a callback or a virtual page in a second argument.

Perhaps adding addTabURL() is better?

**Comment by ibelar** [comment]
Now good to go!

**Comment by romaninsh** [comment]
@ibelar - I have added some cleanups, restored original addTab() functionality and moved url generation outside of the tab. If user does pass the array, then $app->url() is used, but he can call `addTabURL('bleh', $app->jsURL(..))`

I wonder if this implementation is still OK for wordpress admin and if it would still solve #369

**Comment by ibelar** [comment]
@romaninsh - I see, understand the changes.

The changes will still work in WP but we will need to define VirtualPage prior to set Tab and use VirtualPage url like this:

```
        $vp = $this->add(['VirtualPage']);
        $vp->cb->needJsURL = true;
        $vp->add(['LoremIpsum', 'size' => 2]);

        // dynamic tab
        $t->addTabURL('Dynamic Lorem Ipsum', $vp->getURL('cut'));
```

as opposed to just do:

```
        // dynamic tab
        $t->addTabURL('Dynamic Lorem Ipsum', function ($tab) {
            $tab->add(['LoremIpsum', 'size' => 2]);
        }, true);
```

I can also add WpTabs, extending Tabs, in my integration that will take care of this so we can still use it as before.

```
        $t = $this->add('WpTabs');
        // dynamic tab
        $t->addWpTab('Dynamic Lorem Ipsum', function ($tab) {
            $tab->add(['LoremIpsum', 'size' => 2]);
        }, true);
```

Should still solve #369

**Comment by romaninsh** [comment]
Can you explain WHEN VirtualPage does not need "jsURL" ? 

Can we set Callback::needJsURL = true permanently ?

**Comment by ibelar** [comment]
Wizard callback need a regular url for next and previous step buttons. Probably because it does not render json but html only. When pressing next using a link generate via jsUrl, it will only output the form html not the entire page.


May be other in future...

wizard-jsurl

**Comment by romaninsh** [comment]
@ibelar , i have done some thinking and I'm proposing to maintain getURL and getJSURL inside callbacks and virtual pages. Here is my reasoning:

When we create Virtual page or a callback it does not matter what is inside or where it is placed. It actually matters how will we invoke said callback. There are two options:

1. load it dynamically (would need jsURL)
2. direct user to it (would need url)

So technically the same content could be available either directly or through JS. So my proposed solution this time is to update ATK to use 'getJSURL' whenever we specify callback that is used inside JavaScript.

I hope this is better approach. What do you think, @ibelar

**Comment by ibelar** [comment]
Yes! Totally agree with this change. It really make more sense and believe this will also be more flexible for future use.
I have perform some basic testing in WP using the atk-wordpress-demo and atk-demo-sample plugin and all seems to work fine. I am also able to use Wizard out of the box and will be able to include it in atk-wordress-demo.  
Thanks.

**Pull request merged**